### ISSUE TITLE
[reactive-element] Static method finalizeStyles is called even when styles field is not defined

### ISSUE THREAD

**Issue by blikblum** [lit/lit#2881]
# Description

[`finalizeStyles`](https://lit.dev/docs/api/ReactiveElement/#ReactiveElement.finalizeStyles) is called twice when the component class is decorated with a class decorator

This occurs because the decorator sub class the component and the `finalize` static method walks on the prototype chain calling  `finalizeStyles` for each class even if it does not have a `styles` property

# Steps to Reproduce

1. Write this code

```ts
function myClassDecorator(BaseClass) {
  return class extends BaseClass {
    constructor() {
      super()
      console.log('myClassDecorator')    
    }  
  }
}

let finalizeStylesCount = 0

@customElement('simple-greeting')
@myClassDecorator
export class SimpleGreeting extends LitElement {
  static styles = css`p { color: blue }`;
  
  static finalizeStyles(styles) {
    console.count('finalizeStyles')
    finalizeStylesCount++
    return super.finalizeStyles(styles)
  }

  render() {
    return html`<p>finalizeStylesCount: ${finalizeStylesCount}</p>`;
  }
}

```

2. See this output...

finalizeStylesCount: 2

## Live Reproduction Link

[Lit playground](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcbmltcG9ydCB7Y3VzdG9tRWxlbWVudCwgcHJvcGVydHl9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcblxuZnVuY3Rpb24gbXlDbGFzc0RlY29yYXRvcihCYXNlQ2xhc3MpIHtcbiAgcmV0dXJuIGNsYXNzIGV4dGVuZHMgQmFzZUNsYXNzIHtcbiAgICBjb25zdHJ1Y3RvcigpIHtcbiAgICAgIHN1cGVyKClcbiAgICAgIGNvbnNvbGUubG9nKCdteUNsYXNzRGVjb3JhdG9yJykgICAgXG4gICAgfSAgXG4gIH1cbn1cblxubGV0IGZpbmFsaXplU3R5bGVzQ291bnQgPSAwXG5cbkBjdXN0b21FbGVtZW50KCdzaW1wbGUtZ3JlZXRpbmcnKVxuQG15Q2xhc3NEZWNvcmF0b3JcbmV4cG9ydCBjbGFzcyBTaW1wbGVHcmVldGluZyBleHRlbmRzIExpdEVsZW1lbnQge1xuICBzdGF0aWMgc3R5bGVzID0gY3NzYHAgeyBjb2xvcjogYmx1ZSB9YDtcbiAgXG4gIHN0YXRpYyBmaW5hbGl6ZVN0eWxlcyhzdHlsZXMpIHtcbiAgICBjb25zb2xlLmNvdW50KCdmaW5hbGl6ZVN0eWxlcycpXG4gICAgZmluYWxpemVTdHlsZXNDb3VudCsrXG4gICAgcmV0dXJuIHN1cGVyLmZpbmFsaXplU3R5bGVzKHN0eWxlcylcbiAgfVxuXG4gIEBwcm9wZXJ0eSgpXG4gIG5hbWUgPSAnU29tZWJvZHknO1xuXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gaHRtbGA8cD5maW5hbGl6ZVN0eWxlc0NvdW50OiAke2ZpbmFsaXplU3R5bGVzQ291bnR9PC9wPmA7XG4gIH1cbn1cbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9zaW1wbGUtZ3JlZXRpbmcuanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8c2ltcGxlLWdyZWV0aW5nIG5hbWU9XCJXb3JsZFwiPjwvc2ltcGxlLWdyZWV0aW5nPlxuPC9ib2R5PlxuIn0seyJuYW1lIjoicGFja2FnZS5qc29uIiwiY29udGVudCI6IntcbiAgXCJkZXBlbmRlbmNpZXNcIjoge1xuICAgIFwibGl0XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJAbGl0L3JlYWN0aXZlLWVsZW1lbnRcIjogXCJeMS4wLjBcIixcbiAgICBcImxpdC1lbGVtZW50XCI6IFwiXjMuMC4wXCIsXG4gICAgXCJsaXQtaHRtbFwiOiBcIl4yLjAuMFwiXG4gIH1cbn0iLCJoaWRkZW4iOnRydWV9XQ)

## Expected Results

`finalizeStyles` should be called only once when using class decorator

## Actual Results

`finalizeStyles` is called more than once when using class decorator(s)

## Browsers Affected

- [x] Chrome
- [x] Firefox
- [ ] Edge
- [ ] Safari 11
- [ ] Safari 10
- [ ] IE 11

**Comment by justinfagnani** [comment]
This should be no different than subclassing, and not dependent on the decorator. This behaves the same:

```ts
let finalizeStylesCount = 0;

export class BaseClass extends LitElement {
  static styles = css`p { color: blue }`;
  
  static finalizeStyles(styles) {
    console.count('finalizeStyles');
    finalizeStylesCount++;
    return super.finalizeStyles(styles);
  }

  render() {
    return html`<p>finalizeStylesCount: ${finalizeStylesCount}</p>`;
  }
}

@customElement('simple-greeting')
class SimpleGreeting extends BaseClass {
  constructor() {
    super();
    console.log('myClassDecorator');
  }  
}
```

[Playground Link](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcbmltcG9ydCB7Y3VzdG9tRWxlbWVudCwgcHJvcGVydHl9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcblxubGV0IGZpbmFsaXplU3R5bGVzQ291bnQgPSAwO1xuXG5leHBvcnQgY2xhc3MgQmFzZUNsYXNzIGV4dGVuZHMgTGl0RWxlbWVudCB7XG4gIHN0YXRpYyBzdHlsZXMgPSBjc3NgcCB7IGNvbG9yOiBibHVlIH1gO1xuICBcbiAgc3RhdGljIGZpbmFsaXplU3R5bGVzKHN0eWxlcykge1xuICAgIGNvbnNvbGUuY291bnQoJ2ZpbmFsaXplU3R5bGVzJyk7XG4gICAgZmluYWxpemVTdHlsZXNDb3VudCsrO1xuICAgIHJldHVybiBzdXBlci5maW5hbGl6ZVN0eWxlcyhzdHlsZXMpO1xuICB9XG5cbiAgcmVuZGVyKCkge1xuICAgIHJldHVybiBodG1sYDxwPmZpbmFsaXplU3R5bGVzQ291bnQ6ICR7ZmluYWxpemVTdHlsZXNDb3VudH08L3A-YDtcbiAgfVxufVxuXG5AY3VzdG9tRWxlbWVudCgnc2ltcGxlLWdyZWV0aW5nJylcbmNsYXNzIFNpbXBsZUdyZWV0aW5nIGV4dGVuZHMgQmFzZUNsYXNzIHtcbiAgY29uc3RydWN0b3IoKSB7XG4gICAgc3VwZXIoKTtcbiAgICBjb25zb2xlLmxvZygnbXlDbGFzc0RlY29yYXRvcicpO1xuICB9ICBcbn0ifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vc2ltcGxlLWdyZWV0aW5nLmpzXCI-PC9zY3JpcHQ-XG48L2hlYWQ-XG48Ym9keT5cbiAgPHNpbXBsZS1ncmVldGluZyBuYW1lPVwiV29ybGRcIj48L3NpbXBsZS1ncmVldGluZz5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4yLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjEuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl4zLjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMi4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfV0)

And this isn't a bug them: Each class in the hierarchy needs to have its styles finalized so `finalizeStyles` is expected to be called multiple times so that the styles property if overridden can be applied properly for that subclass.

<hr/>

Slightly off-topic, I would recommend that you use the class mixin pattern to augment classes if possible. Mixins generally put the added prototype in the proper place on the prototype chain as the super class of the written class. It's more rare that you want a subclass of the written class.

```ts
function MyMixin(BaseClass) {
  return class extends BaseClass {
    constructor() {
      super();
      console.log('MyMixin');
    }  
  }
}

@customElement('simple-greeting')
export class SimpleGreeting extends MyMixin(LitElement) {
}
```

**Comment by blikblum** [comment]
> This should be no different than subclassing, and not dependent on the decorator.

Fair enough. Using the decorator somehow hides the subclassing but is basically the same and would not be doable to fix only for decorator only.

Still there's some unnecessary work (and memory usage) when the subclass does not declare a `styles` field: the super class styles are finalized twice ending with two (or more) instances of same styles each one set to its corresponding constructor/class. By only calling finalizeStyles when a styles field is declared could save some work and memory, how much depends on the number of subclasses. This would work the same because the subclass would use the finalized super class styles stored in `elementStyles`.

I updated the issue title.

Some context: i'm using the technique described by one of lit devs [here](https://github.com/lit/lit/issues/1977#issuecomment-872645551). By using the code, i was ending with two style elements for each component because a use a class decorator for most of my components. I used a check to this.hasOwnProperty('styles')  in finalizeStyles as a work around and works so far
